### PR TITLE
feat(custom-resources): AwsCustomResource copy physicalResourceId from request when omit it in onUpdate

### LINKED ISSUE #23843: custom-resource: Passthrough physicalResourceId when its not specified in onUpdate at AwsCustomResource
### Describe the feature

Currently, when using AwsCustomResource construct, users have to specify `physicalResourceId` in `onUpdate`. 
This limitation is brought by AwsCustomResource copies `onUpdate` (`physicalResourceId` is NOT mandatory) to `onCreate` (`physicalResourceId` is mandatory) when `onCreate` is not specified.

AwsCustomResource should be able to omit `physicalResourceId` in `onUpdate` with passthrough it from request. This is useful when users want to call an update API without resoure replacing. 

Some AWS APIs send empty response in update. This causes problem in below situation:

- Create API generates unpredictable IDs, and return it in response
  -  Users set `physicalResourceId: PhysicalResourceId.fromResponse()`
- Update API requires resource ID (not Name), and return **empty response**
  - Users provide resouce ID by `new PhysicalResourceIdReference()` in parameters
  - With empty response, users can't use `physicalResourceId: PhysicalResourceId.fromResponse()` or any ID generation

### Use Case

To create / update / delete resources that is not covered by CloudFormation. If API specification matches above conditions, this feature is critical. Like below:

- https://docs.aws.amazon.com/athena/latest/APIReference/API_UpdateNotebook.html
- https://docs.aws.amazon.com/singlesignon/latest/IdentityStoreAPIReference/API_UpdateUser.html

### Proposed Solution

Passthrough `physicalResourceId` from request to response in `onUpdate` when `physicalResourceId` is not specified.
To do that, update input check logic for `physicalResourceId`.

This change is not breaking, because it's not a problem if users can specify `physicalResourceId` in `onUpdate`.

| onCreate | onCreate.physicalResourceId | onUpdate | onUpdate.physicalResourceId | validation error now? | validation error after change? |
| --- | --- | ---| --- | :---: | :---: |
| AwsSdkCall | specified | AwsSdkCall | specified | no | no |
| AwsSdkCall | specified | AwsSdkCall | undefined | yes | **no <-** |
| AwsSdkCall | undefined | AwsSdkCall | undefined | yes | yes |
| AwsSdkCall | undefined | AwsSdkCall | specified | yes | yes |
| undefined (copied from onUpdate) | undefined | AwsSdkCall | specified | no | no |
| undefined (copied from onUpdate) | undefined | AwsSdkCall | undefined | yes | yes |
| AwsSdkCall | specified | undefined | undefined | no | no |
| AwsSdkCall | undefined | undefined | undefined | yes | yes |

### Other Information

Instead of omit `physicalResourceId` to passthrough, we can provide a property like `passthroughPhysicalResourceId: true`. But I think its more intuitive that omit `physicalResourceId`.

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.61.1

### Environment details (OS name and version, etc.)

Node.js 16.13.0, MacOS 12.6.2

### PR DESCRIPTION
**Pull request by konokenj**
AwsCustomResource is now able to omit `physicalResourceId` in `onUpdate` to copy it from request.

Some `UPDATE` AWS APIs responses with an empty body. When users want to call these APIs using AwsCustomResource, users can't specify physicalResourceId by `PhysicalResourceId.fromResponse()`. Furthermore, when the Create API generates an unpredictable ID and this must be passed to the Update API, this Construct could not be used. For example, following APIs match this situation:

- https://docs.aws.amazon.com/athena/latest/APIReference/API_UpdateNotebook.html
- https://docs.aws.amazon.com/singlesignon/latest/IdentityStoreAPIReference/API_UpdateUser.html

Closes #23843.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by comcalvi** [comment]
>AwsCustomResource is now able to omit physicalResourceId in onUpdate with passthrough it from request. This is useful when users want to call an update API without resoure replacing.

I'm not sure what "passthrough it from request" means, could you please elaborate? 

>call an update API without resoure replacing.

I don't understand this either. Does it mean that today, it is impossible to call `onUpdate` without a replacement of the custom resource?

**Comment by konokenj** [comment]
Hi @comcalvi, thank you for response.

> I'm not sure what "passthrough it from request" means, could you please elaborate?

Lambda backed custom resources recieves `physicalResourceId` in request event on update. The handler function has responsibility to return same `physicalResourceId` when the resource is not replaced.

> The value returned for a PhysicalResourceId can change custom resource update operations. If the value returned is the same, it is considered a normal update. If the value returned is different, AWS CloudFormation recognizes the update as a replacement and sends a delete request to the old resource.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/crpg-ref-requesttypes-update.html

But currently, users of AwsCustomResource construct must specify `physicalResourceId` by hand (`fromResponse()` or `.of()`). There is no way to copy from request, although there are AWS APIs that response with empty bodies. This PR adds ability to copy that.

>> call an update API without resoure replacing.
>
> I don't understand this either. Does it mean that today, it is impossible to call onUpdate without a replacement of the custom resource?

No, I intended to allow users to indicate their intention to copy PhysicalResourceId.

(I'm sorry if this is a problem caused by my poor English skill.)

**Review by comcalvi — DISMISSED** [review]
To confirm, leaving the `PhysicalResourceId` undefined will tell CFN to use the value that was passed to`onCreate`? If so, this makes perfect sense. Please create a new unit test in `test/aws-custom-resource` that verifies the updated error checking logic you've added

**Comment by konokenj** [comment]
I added unit tests.

> To confirm, leaving the `PhysicalResourceId` undefined will tell CFN to use the value that was passed to `onCreate`?

`PhysicalResourceId` is not rendered in CFn template when omit in `onUpdate`. When handler function doesn't recieve `PhysicalResourceId` from request, it uses previous value. 
https://github.com/aws/aws-cdk/blob/0ebbf58bdd3307f536334beb5d1153e3ef660f18/packages/%40aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts#L153

**Review by comcalvi — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/custom-resources/test/aws-custom-resource/aws-custom-resource.test.ts [hunk 1]
```diff
@@ -151,40 +151,335 @@ test('with custom policyStatements', () => {
...
   const stack = new cdk.Stack();
-  expect(() => new AwsCustomResource(stack, 'AwsSdk', {
-    policy: AwsCustomResourcePolicy.fromSdkCalls({ resources: AwsCustomResourcePolicy.ANY_RESOURCE }),
-  })).toThrow(/`onCreate`.+`onUpdate`.+`onDelete`/);
+  expect(
+    () =>
+      new AwsCustomResource(stack, 'AwsSdk', {
+        policy: AwsCustomResourcePolicy.fromSdkCalls({ resources: AwsCustomResourcePolicy.ANY_RESOURCE }),
+      }),
+  ).toThrow(/`onCreate`.+`onUpdate`.+`onDelete`/);
```
[the remark this answers is not in the document]
Comment (konokenj) [reply packages/@aws-cdk/custom-resources/test/aws-custom-resource/aws-custom-resource.test.ts]: I reverted style changes.

**Comment by konokenj** [comment]
I fixed code style and error messages. Could you review again?

**Review by comcalvi — APPROVED** [review]

**Pull request merged**